File: 2022/Rust/src/bin/03.rs
aoc2022::main!();

fn generator(input : &str) -> Vec<&str> {
    input.lines().collect()
}

fn part_1(input : Vec<&str>) -> usize {
    solver(input, 1)
}

fn part_2(input : Vec<&str>) -> usize {
    solver(input, 3)
}

fn solver(input : Vec<&str>, size :usize) -> usize {
    input
    .chunks(size)
    .map(|l| 
            l[0].chars().take(l[0].len()/(1 + size % 3)).find(
                |c| 
                (0 .. size).all(
                    |i|
                    l[i].chars().skip((size % 3)*l[0].len()/2).contains(c))
            ).unwrap()
        )
    .map(|c| c as usize - 38 - (c.is_lowercase() as usize) * 58)
    .sum()
}

File: 2018/03/03.py
from parse import parse
from itertools import product

def generator(input):
    return [parse("#{:d} @ {:d},{:d}: {:d}x{:d}", line)[:5] for line in input.splitlines()]

def part_1(input): 
    grid = solver(input)
    return sum(grid[y][x] == -1 for x, y in product(range(len(grid)), range(len(grid[0]))))

def part_2(input): 
    grid = solver(input)
    for n, _, _, w, h in input[::-1]:
        if sum(grid[y][x] == n for x, y in product(range(len(grid)), range(len(grid[0])))) == w * h :
            return n

        
def solver(input):
    grid_size = 1_000
    visited = []
    grid = [[0 for _ in range(grid_size)] for _ in range(grid_size)]
    for n, dx, dy, w, h in input:
        for x, y in product(range(w), range(h)):
            if grid[dy + y][dx + x] == 0: grid[dy + y][dx + x] = n
            else : grid[dy + y][dx + x] = - 1
    return grid